arch: arm: Disallow FP_HARDABI when building with TFM

oyvindronningstad opened the pull request:
When building with TFM, the app is linked with libraries built by the
TFM build system. TFM is always built with -msoft-float which is
equivalent to -mfloat-abi=soft. FP_HARDABI adds -mfloat-abi=hard
which gives errors when linking with the libs from TFM since they are
built with a different ABI.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/33956

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

oyvindronningstad committed:
arch: arm: Disallow FP_HARDABI when building with TFM

When building with TFM, the app is linked with libraries built by the
TFM build system. TFM is always built with -msoft-float which is
equivalent to -mfloat-abi=soft. FP_HARDABI adds -mfloat-abi=hard
which gives errors when linking with the libs from TFM since they are
built with a different ABI.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/33956

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

galak commented:
Do we have the ability to build TFM w/hard float ABI?  Is this a change we need to the TFM build system to allow this?

ioannisg commented:
> Do we have the ability to build TFM w/hard float ABI? Is this a change we need to the TFM build system to allow this?

I am going to investigate this @galak.
In the mean time, I am not sure we can accept dropping the HARD ABI just like that; this does not allow the compiler to use the FP coprocessor efficiently....

tejlmand commented:
> In the mean time, I am not sure we can accept dropping the HARD ABI just like that;

Not sure I follow, hard abi is causing build failures when using TF-M, so we would not be dropping support for anything cause the support has never been there when enabling TF-M.

So this simply ensures users are informed at an early stage.

tejlmand commented on arch/arm/core/aarch32/Kconfig:
```diff
@@ -221,6 +221,7 @@ choice
 
 config FP_HARDABI
 	bool "Floating point Hard ABI"
+	depends on !BUILD_WITH_TFM
```
This is quite a big limitation, could we (perhaps in help text) inform the users why hard fp cannot be used ?

Would also be great if we somewhere (docs ? ) can place a link to any open TF-M issues or tasks around this limitation,

ioannisg replied:
> Not sure I follow, hard abi is causing build failures when using TF-M, so we would not be dropping support for anything cause the support has never been there when enabling TF-M.

I meant that if might be possible to identify what parts of TFM exactly cause the build failures, when using hard ABI in Zephyr, isolate them, and offer hard ABI support with TFM, if these parts are not used. That's my idea at least - not sure about how practical this is.

tejlmand replied:
Got it, that would indeed be nice if it were possible, but looking at the top-level CMakeLists.txt file in TF-M, that's gonna be hard:
https://github.com/zephyrproject-rtos/trusted-firmware-m/blob/d07d72c64e4b3d8594829eea302b88e51709861f/trusted-firmware-m/CMakeLists.txt#L29-L31
```
if(TFM_SYSTEM_FP)
    message(FATAL_ERROR "Hardware FPU is currently not supported in TF-M")
endif()
```

ioannisg replied:
@tejlmand a comment was added to explain the limitation.

ioannisg committed:
arm: add note explaining why Hard ABI is disabled for tfm builds

Add a note in the Kconfig help text that explains why Hard ABI
is not possible on builds with TF-M.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg reviewed (approved):
This is as good as it can be - unfortunately, I have not found a reasonable way to get around without this fix.
The only case where we could have TFM compiled with SOFT ABI while Zephyr as Hard ABI is when the binaries would not need to be linked. That's the case where TFM would only need to do a secure boot, but would contain no secure service calls. This use case is perhaps not practical today, so I am fine with accepting the patch that fully disallows hard ABI on builds with TFM.

Eventually, this is something that TFM needs to fix.

PerMac reviewed (approved):
I've rerun the tests reported in ##33956 and now they are passing
